packages/contracts/contracts/Router.sol

```diff
@@ -0,0 +1,220 @@
+    address assetId;
+    uint256 chainId; // For domain separation
+  }
+
+  event RelayerFeeAdded(address assetId, uint256 amount, address caller);
+  event RelayerFeeRemoved(address assetId, uint256 amount, address caller);
+
+  constructor(address _routerFactory) {
+    routerFactory = _routerFactory;
+  }
+
+  // Prevents from calling methods other than routerFactory contract
+  modifier onlyViaFactory() {
+    require(msg.sender == routerFactory, "ONLY_VIA_FACTORY");
+    _;
+  }
+
+  function init(
+    address _transactionManager,
+    uint256 _chainId,
+    address _routerSigner,
+    address _recipient,
+    address _owner
+  ) external onlyViaFactory {
+    transactionManager = ITransactionManager(_transactionManager);
+    chainId = _chainId;
+    routerSigner = _routerSigner;
+    recipient = _recipient;
+    transferOwnership(_owner);
+  }
+
+  function setRecipient(address _recipient) external onlyOwner {
+    recipient = _recipient;
+  }
+
+  function setSigner(address _routerSigner) external onlyOwner {
+    routerSigner = _routerSigner;
+  }
+
+  function addRelayerFee(uint256 amount, address assetId) external payable {
+    // Sanity check: nonzero amounts
+    require(amount > 0, "#RC_ARF:002");
+
+    // Transfer funds to contract
+    // Validate correct amounts are transferred
+    if (LibAsset.isNativeAsset(assetId)) {
+      require(msg.value == amount, "#RC_ARF:005");
+    } else {
+      require(msg.value == 0, "#RC_ARF:006");
+      LibAsset.transferFromERC20(assetId, msg.sender, address(this), amount);
+    }
+
+    // Emit event
+    emit RelayerFeeAdded(assetId, amount, msg.sender);
+  }
+
+  function removeRelayerFee(uint256 amount, address assetId) external onlyOwner {
+    // Sanity check: nonzero amounts
+    require(amount > 0, "#RC_RRF:002");
+
+    // Transfer funds from contract
+    LibAsset.transferAsset(assetId, payable(recipient), amount);
+
+    // Emit event
+    emit RelayerFeeRemoved(assetId, amount, msg.sender);
+  }
+
+  function removeLiquidity(
+    uint256 amount,
+    address assetId,
+    bytes calldata signature
+  ) external {
+    if (msg.sender != routerSigner) {
+      SignedRemoveLiquidityData memory payload = SignedRemoveLiquidityData({
+        amount: amount,
+        assetId: assetId,
+        chainId: chainId
+      });
+
+      address recovered = recoverSignature(abi.encode(payload), signature);
+      require(recovered == routerSigner, "#RC_RL:040");
+    }
+
+    return transactionManager.removeLiquidity(amount, assetId, payable(recipient));
+  }
+
+  function prepare(
+    ITransactionManager.PrepareArgs calldata args,
+    address routerRelayerFeeAsset,
+    uint256 routerRelayerFee,
+    bytes calldata signature
+  ) external payable returns (ITransactionManager.TransactionData memory) {
+    if (msg.sender != routerSigner) {
+      SignedPrepareData memory payload = SignedPrepareData({
+        args: args,
+        routerRelayerFeeAsset: routerRelayerFeeAsset,
```

Comment:
Won't the relayer fee *always* be in receiving token?

Reply:
in the router contract you could specify what asset and just park fees on the contract directly. the relayer fee from the transaction manager would actually be sent to this contract, so we should make sure that we can easily handle that case